README.md

<!-- README.md is generated from README.Rmd. Please edit that file -->

# dblp

<!-- badges: start -->

[![Lifecycle:
experimental](https://img.shields.io/badge/lifecycle-experimental-orange.svg)](https://www.tidyverse.org/lifecycle/#experimental)
<!-- badges: end -->

The dblp computer science bibliography at [dblp.org](https://dblp.org)
provides open bibliographic information on major computer science
journals and proceedings. Originally created at the University of Trier
in 1993, dblp is now operated and further developed by Schloss Dagstuhl.
This R package interfaces with the API, making data available to use
from R.

## Installation

You can install the development version of dblp from GitHub with:

``` r
#install.packages("devtools")
install_github("KTH-Library/dblp", dependencies = TRUE)
```

## Example

This is a basic example which shows you how make a search :

``` r
library(dblp)
library(knitr)
library(stringr)
suppressPackageStartupMessages(library(dplyr))

# search for publications

dblp_search("quantum computer")$content %>%
  select(starts_with("info")) %>%
  mutate(result = paste(`info.title`, `info.ee`, `info.url`)) %>%
  select(result) %>%
  slice(1:5) %>%
  kable()
```

| result                                                                                                                                                                                                                                               |
| :--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- |
| Implementation of quantum secret sharing and quantum binary voting protocol in the IBM quantum computer. <https://doi.org/10.1007/s11128-019-2531-z> <https://dblp.org/rec/journals/qip/JoySBP20>                                                    |
| Quantum Computing for Computer Architects, Second Edition <https://doi.org/10.2200/S00331ED1V01Y201101CAC013> <https://dblp.org/rec/series/synthesis/2011Metodi>                                                                                     |
| Quantum Communication and Quantum Networking, First International Conference, QuantumComm 2009, Naples, Italy, October 26-30, 2009, Revised Selected Papers <https://doi.org/10.1007/978-3-642-11731-2> <https://dblp.org/rec/conf/quantumcomm/2009> |
| Quantum Computer Science <https://doi.org/10.2200/S00159ED1V01Y200810QMC002> <https://dblp.org/rec/series/synthesis/2008Lanzagorta>                                                                                                                  |
| Quantum Walks for Computer Scientists <https://doi.org/10.2200/S00144ED1V01Y200808QMC001> <https://dblp.org/rec/series/synthesis/2008Venegas>                                                                                                        |

``` r

# crawl when there are many pages of results


search <- dblp_crawl("Royal Institute")$content

# display results
search %>%
  filter(str_detect(`info.ee`, "kth:diva")) %>%
  arrange(desc(`info.year`)) %>%
  # rename and order some named and indexed columns
  select(DOI = `info.doi`, 9:4) %>%
  # add a HTML link by combining title and use DOI for href
  mutate(link = sprintf("<a href='https://doi.org/%s'>%s...</a>", 
    DOI, str_sub(`info.title`, 1, 50))) %>%
  # pick out a few columns to display in a HTML table
  select(link, 1:4, -c("DOI", "info.ee"), `info.year`) %>%
  slice(1:5) %>%
  kable()
```

| link                                                                                 | info.url                                           | info.key                    | info.year |
| :----------------------------------------------------------------------------------- | :------------------------------------------------- | :-------------------------- | :-------- |
| <a href='https://doi.org/NA'>Data-driven Methods in Inverse Problems….</a>           | <https://dblp.org/rec/phd/basesearch/Adler19>      | phd/basesearch/Adler19      | 2019      |
| <a href='https://doi.org/NA'>Scalable Analysis of Large Datasets in Life Scienc…</a> | <https://dblp.org/rec/phd/basesearch/Ahmed19>      | phd/basesearch/Ahmed19      | 2019      |
| <a href='https://doi.org/NA'>Feasibility and Performance of Dynamic TDD in Dens…</a> | <https://dblp.org/rec/phd/basesearch/Celik19>      | phd/basesearch/Celik19      | 2019      |
| <a href='https://doi.org/NA'>Graph Algorithms for Large-Scale and Dynamic Natur…</a> | <https://dblp.org/rec/phd/basesearch/Ghoorchian19> | phd/basesearch/Ghoorchian19 | 2019      |
| <a href='https://doi.org/NA'>Macroscopic models of Chinese hamster ovary cell c…</a> | <https://dblp.org/rec/phd/basesearch/Hagrot19>     | phd/basesearch/Hagrot19     | 2019      |

``` r

# search for venues

dblp_crawl("Europe", entity = "venues")$content %>%
  slice(1:5) %>%
  select(starts_with("info")) %>%
  kable()
```

| info.venue                                                                        | info.acronym | info.type              | info.url                              |
| :-------------------------------------------------------------------------------- | :----------- | :--------------------- | :------------------------------------ |
| Central European Cybersecurity Conference (CECC)                                  | CECC         | Conference or Workshop | <https://dblp.org/db/conf/cecc/>      |
| Central European Functional Programming School (CEFP)                             | CEFP         | Conference or Workshop | <https://dblp.org/db/conf/cefp/>      |
| Central European Journal of Operations Research                                   | NA           | Journal                | <https://dblp.org/db/journals/cejor/> |
| Central and East European Conference on Software Engineering Techniques (CEE-SET) | CEE-SET      | Conference or Workshop | <https://dblp.org/db/conf/ifip2/>     |
| Central-European Workshop on Services and their Composition (ZEUS)                | ZEUS         | Conference or Workshop | <https://dblp.org/db/conf/zeus/>      |

``` r

# search for authors

dblp_crawl("maguire", entity = "authors")$content %>%
  select(starts_with("info")) %>%
  filter(stringr::str_starts(`info.author`, "Ger")) %>%
  kable()
```

| info.author           | info.url                             | info.aliases.alias | info.notes.note.@type | info.notes.note.text                                 |
| :-------------------- | :----------------------------------- | :----------------- | :-------------------- | :--------------------------------------------------- |
| Gerald Q. Maguire Jr. | <https://dblp.org/pid/m/GQMaguireJr> | NULL               | affiliation           | KTH Royal Institute of Technology, Stockholm, Sweden |
| Gerry Maguire         | <https://dblp.org/pid/255/3148>      | NULL               | NA                    | NA                                                   |


NEWS.md
# dblp 0.0.0.9000

* Added a `NEWS.md` file to track changes to the package.


tests/testthat.R
library(testthat)
library(dblp)

test_check("dblp")


tests/testthat/test-crawl.R
test_that("crawl works", {
  t1 <- dblp_crawl("chips", entity = "publications")$content
  expect_gt(nrow(t1), 1000)
})


tests/testthat/test-search.R
test_that("search works", {
  t1 <- dblp_search("European", entity = "venues")$content
  expect_gt(nrow(t1), 5)
})

test_that("search venues w zero hits works", {
  t1 <- dblp_search("venues", entity = "venues")$content
  expect_equal(nrow(t1), 0)
})

test_that("search authors w zero hits works", {
  t1 <- dblp_search("nooneinparticular", entity = "authors")$content
  expect_equal(nrow(t1), 0)
})


R/utils.R
#' @importFrom attempt stop_if_not
#' @importFrom curl has_internet
check_internet <- function(){
  stop_if_not(.x = has_internet(), msg = "Please check your internet connexion")
}

#' @importFrom httr status_code
check_status <- function(res){
  stop_if_not(
    .x = status_code(res),
    .p = ~ .x == 200,
    msg = "The API returned an error")
}

#' dblp API configuration
#'
#' This function provides a list with settings for the base url and user agent
#' used in API calls.
#'
#' @export
config <- function() {
  list(
    url_publications = "http://dblp.org/search/publ/api",  # for publication queries
    url_authors = "http://dblp.org/search/author/api",  # for author queries
    url_venues = "http://dblp.org/search/venue/api", # for venue queries
    ua = httr::user_agent("http://github.com/hadley/httr")
  )
}



data-raw/dev.R
# https://www.r-bloggers.com/how-to-build-an-api-wrapper-package-in-10-minutes/
# https://httr.r-lib.org/articles/api-packages.html

install.packages("devtools")
install.packages("roxygen2")
install.packages("usethis")
install.packages("curl")
install.packages("httr")
install.packages("jsonlite")
install.packages("attempt")
install.packages("purrr")
devtools::install_github("r-lib/desc")

library(devtools)
library(usethis)
library(desc)

unlink("DESCRIPTION")
my_desc <- description$new("!new")
my_desc$set("Package", "dblp")
my_desc$set("Authors@R", "person('Markus', 'Skyttner', email = 'markussk@kth.se', role = c('cre', 'aut'))")

my_desc$del("Maintainer")

my_desc$set_version("0.0.0.9000")

my_desc$set(Title = "Computer Science Bibliographic Data from the dblp.org API")
my_desc$set(Description = "The dblp computer science bibliography at dblp.org provides open bibliographic information on major computer science journals and proceedings. Originally created at the University of Trier in 1993, dblp is now operated and further developed by Schloss Dagstuhl. This R package interfaces with the API, making data available to use from R.")
my_desc$set("URL", "https://github.com/KTH-Library/dblp")
my_desc$set("BugReports", "https://github.com/KTH-Library/dblp/issues")
my_desc$set("License", "MIT")
my_desc$write(file = "DESCRIPTION")

use_mit_license(name = "Markus Skyttner")
#use_code_of_conduct()
use_news_md()
use_readme_rmd()
use_lifecycle_badge("Experimental")

use_package("httr")
use_package("jsonlite")
use_package("curl")
use_package("attempt")
use_package("purrr")
use_package("progress")
use_package("dplyr")
use_package("tibble")
use_package("rlang")
use_package("stringr")
use_tidy_description()

use_testthat()
use_vignette("dblp")
use_roxygen_md()
use_pkgdown()

document()
test()
check()
pkgdown::build_site()


R/dblp.R


#' Retrieve records from dblp API
#'
#' The database used can be specified - the default is to use LIBRIS but "swepub" is also available
#'
#' @param query the query string to search for, as described in \url{https://dblp.org/faq/13501473.html}, for example q=test+search
#' @param start the "f"irst hit in the numbered sequence of search results to return. In combination with the h parameter, this parameter can be used for pagination of search results
#' @param n maximum number of search results ("h"its) to return. For bandwidth reasons, this number is capped at 1000
#' @param c maximum number of completion terms to return. For bandwidth reasons, this number is capped at 1000
#' @param api_config a configuration setting for the API including base URL etc, by default from config()
#' @param entity one of "publications", "authors" or "venues" where "publications" is default
#' @importFrom attempt stop_if_all stop_if_not
#' @importFrom purrr compact
#' @importFrom jsonlite fromJSON
#' @importFrom httr GET http_type status_code
#' @importFrom progress progress_bar
#' @import tibble dplyr
#' @export
#'
#' @return results records returned from the search
#' @examples
#' \dontrun{
#' dblp_search()
#' }
#'
#'
dblp_search <- function(
  query = NULL, start = 1, n = 10, c = NULL,
  api_config = NULL, entity = c("publications", "authors", "venues"))
{
  query_args <- list(
    q = query, format = "json", f = start, h = n
  )

  check_internet()
  stop_if_all(args, is.null, "You need to specify at least one argument")

  stop_if_not(is.character(query) && nchar(query) > 0,
              msg = "Please provide a query")
  stop_if_not(n <= 1000, msg = "Maximum is 1000 for number of returned records")

  if (any(missing(api_config), is.null(api_config)))
    api_config <- config()

  path <- switch(match.arg(entity),
    "publications" = api_config$url_publications,
    "authors" = api_config$url_authors,
    "venues" = api_config$url_venues
  )

  resp <- GET(path, query = compact(query_args), api_config$ua)
  check_status(resp)
  if (http_type(resp) != "application/json") {
    stop("API did not return json", call. = FALSE)
  }

  parsed <- fromJSON(rawToChar(resp$content))

  if (status_code(resp) != 200) {
    stop(
      sprintf(
        "API request failed [%s]\n%s\n<%s>",
        status_code(resp),
        parsed$message,
        parsed$documentation_url
      ),
      call. = FALSE
    )
  }

  # TODO - parse nested json into tabular format, for now just remove compound values
  parse_publications <- function(x) {
    tryCatch(
      x$result$hits$hit %>%
      jsonlite::flatten() %>%
      as_tibble(), #as_tibble(.name_repair = "universal")
      error = function(e) as_tibble(NULL)
    )
  }

  parse_authors <- parse_publications
  parse_venues <- parse_publications

  content <- switch(match.arg(entity),
    "publications" = parse_publications(parsed),
    "authors" = parse_authors(parsed),
    "venues" = parse_venues(parsed)
  )

  structure(
    list(
      content = content,
      query = query,
      from = parsed$result$hits$`@first` %>% as.integer(),
      to = parsed$result$hits$`@sent` %>% as.integer(),
      count = parsed$result$hits$`@total` %>% as.integer()
    ),
    class = "dblp"
  )
}

#' Retrieve more records from dblp in batches
#'
#' This function retrieves batches of record (paging through results)
#' @inheritDotParams dblp_search
#' @export
#' @return results records returned from the search
#' @importFrom rlang eval_tidy parse_quo current_env
#' @examples
#' \dontrun{
#' dblp_crawl(query = "chip")
#' }
dblp_crawl <- function(...) {
  res <- dblp_search(...)
  n <- res$count

  if (n <= 10)
    return (res)

  batch <- 1000L

  if (n > 10 & n < batch) {
    res <- dblp_search(..., n = n)
    return (res)
  }

  if (n >= batch & n < 10000) {
    # nr of pages to fetch
    n_pages <- (n %/% batch) + ifelse(n %% batch > 0, 1, 0)

    i <- c(0, (1:(n_pages - 1)) * batch)

    calls <- sprintf("dblp_search(start = %s, n = %s, ...)$content", i, batch)
    message(sprintf("Fetching %s hits in %s batches of %s records", n, n_pages, batch))

    pb <- progress::progress_bar$new(
      format = "  downloading [:bar] :percent in :elapsed",
      total = n_pages, clear = FALSE, width= 60)

    call2df <- function(x) {
      pb$tick()
      Sys.sleep(0.01)
      force(rlang::eval_tidy(rlang::parse_quo(x, env = rlang::current_env()))) #%>%
      #as.data.frame() %>%
      #tibble::as_tibble() %>%
      #dplyr::select_if(function(y) !is.list(y))
    }

    batch <- calls %>% purrr::map_df(call2df)

    res <- structure(
      list(
        content = batch,
        query = sprintf("%s pages of data", n_pages),
        from = 1,
        to = n + 1,
        count = n
      ),
      class = "dblp"
    )
  } else {
    stop(sprintf("Too big batch, %s records, stopping", n), call. = FALSE)
  }

  return(res)
}

#' @export
print.dblp <- function(x, ...) {
  stopifnot(inherits(x, 'dblp'))
  cat(sprintf("<dblp.org search for '%s' (n_hits=%s, showing %s..%s)>\n",
              x$query, x$count, x$from, x$to))
  print(tibble::as_tibble(x$content))
  invisible(x)
}
